### PR TITLE
Introduce sidecar tensorboard

### PR DESCRIPTION
**Pull request by zuston**

### CODE REVIEW

File: tony-core/src/main/java/com/linkedin/tony/runtime/MLGenericRuntime.java [hunk 1]
```diff
@@ -92,9 +114,29 @@ public boolean validateAndUpdateConfig(Configuration tonyConf) {
...
+    private void checkAndPrepareTBResource(Configuration tonyConf) {
+        if (!enableSidecarTB(tonyConf)) {
+            return;
+        }
+
+        // TODO: 2021/5/8 check illegal keys. [tensorboard] role is illegal
+
+        tonyConf.set(TENSORBOARD_RESOURCE_VCORE, "1");
+        tonyConf.set(TENSORBOARD_RESOURCE_INSTANCES, "1");
+        tonyConf.set(TENSORBOARD_RESOURCE_MEMORY, "2g");
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/runtime/MLGenericRuntime.java:134]: create default values in configuration (like TENSORBOARD_RESOURCE_INSTANCES_DEFAULT)

File: tony-core/src/main/java/com/linkedin/tony/runtime/MLGenericRuntime.java [hunk 2]
```diff
@@ -142,12 +184,77 @@ protected void printTasksPeriodically() {
...
+        assert taskExecutor != null;
+
+        boolean enableSidecarTB = enableSidecarTB(taskExecutor.getTonyConf());
+        if (!enableSidecarTB && taskExecutor.isChief()) {
+            return true;
+        }
+
+        if (enableSidecarTB && isSidecarTBExecutor(taskExecutor)) {
+            return true;
+        }
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/runtime/MLGenericRuntime.java:202]: add some comments for the logic

**Review by oliverhu — COMMENTED** [review]
overall looks good

File: tony-core/src/main/java/com/linkedin/tony/runtime/MLGenericRuntime.java [hunk 3]
```diff
@@ -142,12 +188,85 @@ protected void printTasksPeriodically() {
...
+        buildTaskEnv(taskExecutor);
+        return executorPythonShell(taskExecutor);
+    }
+
+    private void setStartupTBResource() {
+        String tbLogDir = taskExecutor.getTonyConf().get(TonyConfigurationKeys.TENSORBOARD_LOG_DIR);
+        Path tbScriptPath = createSidecarTBScript();
+        String pythonExecPath = taskExecutor.getTonyConf().get(TonyConfigurationKeys.PYTHON_EXEC_PATH, "python");
+        String startTBCommand = String.format("%s %s --logdir %s --port %s",
+                pythonExecPath, tbScriptPath, tbLogDir, taskExecutor.getTbPort());
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/runtime/MLGenericRuntime.java:238]: how is this registered with AM trackingURL?

Comment (zuston) [reply tony-core/src/main/java/com/linkedin/tony/runtime/MLGenericRuntime.java:238]: Here. https://github.com/linkedin/TonY/blob/9010e0272923a51698954139bab68d47c1dcf69d/tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java#L98

**Review by oliverhu — APPROVED** [review]

**Pull request merged**